Kotlin in-memory compiler support.

brettwooldridge opened the pull request:
Changes to support #1037, #956, and #395.

kageiit commented on src/com/facebook/buck/android/KotlinAndroidLibraryCompiler.java:
```diff
@@ -43,8 +43,18 @@ public CompileToJarStepFactory compileToJar(
         ANDROID_CLASSPATH_FROM_CONTEXT);
   }
+
+//  @Override
```
Can we clean up these comments?

kageiit commented on buck.iml:
```diff
@@ -92,5 +92,8 @@
     <orderEntry type="library" name="jna-4.2.0" level="project" />
     <orderEntry type="library" name="jna-platform-4.2.0" level="project" />
     <orderEntry type="library" name="rocksdbjni-5.1.2" level="project" />
+    <orderEntry type="library" name="kotlin-compiler-1.1" level="project" />
```
Considering kotlin versions evolve rapidly, it may not make sense to bundle kotlin jars with buck for every release. Maybe its best if we let consumers point to the kotlin jars they want as `remote_file`s or jars etc.?

brettwooldridge replied:
@kageiit This can be removed, because everything is done via reflection now.

kageiit commented on src/com/facebook/buck/jvm/kotlin/BUCK:
```diff
@@ -1,9 +1,22 @@
+    immutable_types = [
+        'KotlincVersion',
+    ],
+    autodeps = True,
```
I think autodeps was mostly removed by @Coneko recently

kageiit commented on src/com/facebook/buck/jvm/kotlin/DefaultKotlinLibraryBuilder.java:
```diff
@@ -53,9 +55,42 @@ protected BuilderHelper newHelper() {
+
+    @Override
     protected CompileToJarStepFactory buildCompileStepFactory() {
+//      new KotlincToJarStepFactory(
```
cleanup comments

kageiit commented on src/com/facebook/buck/jvm/kotlin/JarBackedReflectedKotlinc.java:
```diff
@@ -0,0 +1,237 @@
+      Class<?> exitCodeClass = compilerShim
+          .getClass()
+          .getClassLoader()
+          .loadClass("org.jetbrains.kotlin.cli.common.ExitCode");
```
move this to a constant

kageiit replied:
We can also cache the class obtained this way to avoid needing to load the class for every compile invocation

brettwooldridge replied:
Java reflection already performs caching, a subsequent call to ``loadClass()`` is basically a hash lookup.

kageiit commented on src/com/facebook/buck/jvm/kotlin/JarBackedReflectedKotlinc.java:
```diff
@@ -0,0 +1,237 @@
+            .map(PATH_TO_URL)
+            .collect(Collectors.toList())));
+
+      return classLoader.loadClass("org.jetbrains.kotlin.cli.jvm.K2JVMCompiler").newInstance();
```
constat

kageiit commented on src/com/facebook/buck/jvm/kotlin/KotlinLibraryDescription.java:
```diff
@@ -129,9 +129,25 @@ public Arg createUnpopulatedConstructorArg() {
+    ImmutableList<String> extraKotlincArguments = ImmutableList.of();
   }
+
+//  private ImmutableSortedSet<BuildRule> buildCompileTimeClasspathDeps(
```
cleanup comments

kageiit commented on src/com/facebook/buck/jvm/kotlin/KotlinTestDescription.java:
```diff
@@ -90,7 +85,15 @@ public Arg createUnpopulatedConstructorArg() {
-    JavaLibrary testsLibrary = resolver.addToIndex(defaultJavaLibraryBuilder.build());
+    DefaultKotlinLibrary testsLibrary = (DefaultKotlinLibrary) resolver.addToIndex(defaultJavaLibraryBuilder.build());
+
+//    Kotlinc kotlinc = kotlinBuckConfig.getKotlinc();
```
cleanup comments

runningcode commented on src/com/facebook/buck/jvm/kotlin/DefaultKotlinLibrary.java:
```diff
@@ -0,0 +1,84 @@
+import java.util.Set;
+import java.util.regex.Pattern;
+
+public class DefaultKotlinLibrary extends DefaultJavaLibrary {
```
Is there a difference between this class and DefaultJavaLibrary? I'm new to this codebase and this seems like a common pattern here. I'm just thinking we should document things like this. :)

runningcode commented on src/com/facebook/buck/jvm/kotlin/ExternalKotlinc.java:
```diff
@@ -0,0 +1,174 @@
+
+public class ExternalKotlinc implements Kotlinc {
+
+  private static final KotlincVersion DEFAULT_VERSION = KotlincVersion.of("unknown version");
```
is this the default version on unknown version?
can we document the behavior around versions somewhere? how to set the version, what the default is, what happens if none is set?

brettwooldridge commented:
@kageiit I should have been clear.  This contains changes *toward* support of kapt, i.e. it lays the foundation for completion of #956.  We now invoke the ``K2JVMCompiler.exec()`` method, which winds up [here](https://github.com/JetBrains/kotlin/blob/master/compiler/cli/src/org/jetbrains/kotlin/cli/jvm/K2JVMCompiler.kt#L70) in the CLICompiler.  It is a short distance to go to thread the correct configuration through from buck.

brettwooldridge commented:
@kageiit Just to be clear, this pull request was not made because the code is ready to merge, it was made due to @dsyang request:
> Would you be comfortable putting your changes in kotlin-compiler up in a PR so we could start reviewing it?
>
> I know there's still some work that you'd like to do but this could make it easier to merge later down the line.

kageiit commented:
> this pull request was not made because the code is ready to merge

Makes sense. I left a few comments in places but I guess there still needs to be clean up before merging. Looking forward to this being part of buck! :)

dsyang commented:
Thanks for putting this up @brettwooldridge! Main reason I suggested it is that it's one step closer to merging (and merging means no more rebases from hell).

kageiit commented:
@dsyang @brettwooldridge can we try to merge this or bits of it maybe? It makes it lot easier to do incremental changes on top then

brettwooldridge commented:
@kageiit @dsyang I keep rebasing it once a week or so.  I've got a tiny bit of cleanup now that I realized that it was not possible for an entry in the ``.buckconfig`` to reference a ``prebuilt_jar`` rule.  Give me 'til Monday or so before doing any hasty merging.

The good news is that our development team has been using this branch as our primary *buck* and compiling lots of Kotlin code for well over a month.

dsyang commented on test/com/facebook/buck/jvm/kotlin/KotlinTestAssumptions.java:
```diff
@@ -32,7 +32,7 @@ public static void assumeUnixLike() {
     Throwable exception = null;
     try {
-      new KotlinBuckConfig(FakeBuckConfig.builder().build()).getKotlinCompiler();
+      new KotlinBuckConfig(FakeBuckConfig.builder().build()).getPathToCompilerJar(ruleFinder);
```
ruleFinder doesn't exist.

dsyang commented:
Google just announced kotlin as a official language for Android. @brettwooldridge any chance we can get this PR updated and ready soon? ;)

dsyang commented on src/com/facebook/buck/android/AndroidLibraryCompilerFactory.java:
```diff
@@ -26,5 +26,6 @@
-  AndroidLibraryCompiler getCompiler(JvmLanguage language);
+  AndroidLibraryCompiler getCompiler(
+      JvmLanguage language);
 }
```
unrelated change

dsyang commented on src/com/facebook/buck/android/DefaultAndroidLibraryCompilerFactory.java:
```diff
@@ -34,7 +34,8 @@ public DefaultAndroidLibraryCompilerFactory(
   @Override
-  public AndroidLibraryCompiler getCompiler(AndroidLibraryDescription.JvmLanguage language) {
+  public AndroidLibraryCompiler getCompiler(
+      AndroidLibraryDescription.JvmLanguage language) {
```
unrelated change. please revert.

dsyang commented on src/com/facebook/buck/android/KotlinAndroidLibraryCompiler.java:
```diff
@@ -27,7 +27,8 @@
 
-  public KotlinAndroidLibraryCompiler(KotlinBuckConfig kotlinBuckConfig) {
+  public KotlinAndroidLibraryCompiler(
+      KotlinBuckConfig kotlinBuckConfig) {
```
unrelated change

dsyang commented on src/com/facebook/buck/jvm/kotlin/AbstractKotlincVersion.java:
```diff
@@ -0,0 +1,33 @@
+
+@Value.Immutable
+@BuckStyleImmutable
+public abstract class AbstractKotlincVersion {
```
javadoc

dsyang replied:
https://travis-ci.org/facebook/buck/jobs/235076076#L4351

dsyang commented on src/com/facebook/buck/jvm/kotlin/DefaultKotlinLibraryBuilder.java:
```diff
@@ -52,9 +61,49 @@ protected BuilderHelper newHelper() {
+      }
       return new KotlincToJarStepFactory(
-          Preconditions.checkNotNull(kotlinBuckConfig).getKotlinCompiler().get(),
+          Preconditions.checkNotNull(kotlinBuckConfig).getKotlinc(),
```
can you use the above `kotlinc` here?

Coneko commented:
You can run https://github.com/google/google-java-format on the files to format them automatically and remove spurious unrelated changes.

brettwooldridge commented:
I'll take a look.  I haven't gotten around to minimizing the diffs. Just working on the unit tests.  Once they're cleaned up and running I'll minimize the changes and the pull should be ready.

dsyang commented:
@brettwooldridge sounds good! Anything I can do to help? Happy to patch this PR and help test & fix unit tests.

brettwooldridge commented:
@dsyang I've rebased my changed on to master, but I'm getting a difficult to trace exception from the Kotlin compiler.  Working forward in steps from my April 27 commit, if I rebased onto a May 1 commit then everthing is still working, but if I rebase the same changed onto a May 10 commit I get the same failure.

It's 3:30am here, so I'm going to knock off for tonight.  Tomorrow I'll see if I can use git bisect to narrow down the issue, or do a manual bisect.  Seems classloader related, but I wonly't know until I actually find the breaking commit.  If I can resolve the issue on that commit, I suspect the rebase from there to the head of master will go smoothly.

dsyang commented:
Gotcha. I'll try and narrow it down in https://github.com/dsyang/buck/tree/kotlin-compiler2

Is the weird bug you're seeing this? https://gist.github.com/dsyang/9d5b7562ad676a10b18adc6d427cffda that's what I see when trying to use this branch.  

Thanks for all your hard work! Hope to have some good news for you tomorrow morning.

dsyang commented:
Not sure why I can't submit PRs to @brettwooldridge's buck repo but here's what I found:
 - there was a [trivial compilation error](https://github.com/dsyang/buck/commit/1c573c55d5fab90a118aac17d0271c85cfecb497) that prevented the kotlin tests from running. After fixing that tests would run but I would get a weird kotlin compiler exception ([stacktrace](https://gist.github.com/dsyang/9d5b7562ad676a10b18adc6d427cffda))
 - I traced this error to be related to the v1.1.x kotlin compilers.  If I tell the tests to use v1.0.7 by adding a .buckconfig like below, then all the tests pass:
```
# filename: test/com/facebook/buck/jvm/kotlin/testdata/kotlin_(library|test)_description/.buckconfig
[kotlin]
  compiler = "/Users/dsyang/Downloads/kotlinc-1.0.7/bin/kotlinc"
```
 - If I change the test's`.buckconfig` to use the compiler_jar, I run into another crash ([stacktrace](https://gist.github.com/dsyang/c966a74e5b47b1088c91342ec018b6fc)).  @brettwooldridge  was this the crash you were talking about?

The last crash is more tractable than the first one because it seems to be in the buck codebase.  I'll try and tackle that later today.

brettwooldridge commented:
@dsyang Yes, that is the error.

I have been running my ``kotlin`` branch (literally that is the name) with Kotlin v1.1.1 for some time without that error.  That is the branch from April 27.  When I rebase/merge that into master at May 10th, it fails with that error.  If I merge a bit further back, May 1, it still succeeds.  I feel like it is classloader related (the JarBackedKotlinCompiler uses a distinct classloader).


I have my rebased changes and cleanup on a branch called ``kotlin-merge`` -- failing with the same error.

brettwooldridge commented:
@dsyang Here's a thought.  There are significant changes in ``kotlin-merge``.  In Kotlin v1.1.1 it is no longer sufficient to specify a compiler jar and runtime jar, the kotlin compiler does it's own jar discovery based off of the location of the compiler jar, and pulls in *other* Kotlin jars.

I am wondering if the ``mapdb`` thirdparty component is somehow pulling in the thirdparty/java/kotlin jars -- which may be incompatible with the compiler.  Those are no longer needed or used by the Kotlin integration.

The Kotlin compiler integration required a true installation of Kotlin -- at least it needs all of the jars to be present.  The only .buckconfig variables are now ``kotlin_home`` and ``external`` (true for kotlinc shell out, false for in-process compilation).

dsyang commented:
Looking at `kotlin-merge` now. I think third-party/java/kotlin is unnecessary for `mapdb` to work.

Agreed that it looks like a true installation of kotlin is necessary.  I think the .buckconfig variables at `runtime_jar`, `compiler_jar`, and `compiler`, `KotlinBuckConfig` only checks `kotlin_home` as a environment variable.  Maybe we should replace `runtime_jar` and `compiler_jar` with `kotlin_home`.

brettwooldridge commented:
The KotlinBuckConfig in the kotlin-merge branch only looks for *kotlin_home* and *external*.  I'm pretty sure that the mapdb dependency is the cause.

dsyang commented:
Do you have a theory about why it would be causing this to fail?

brettwooldridge commented:
I'm away from my computer now, but maybe it is shading some Kotlin classes?  Maybe I shouldn't speculate... I'll have a chance to look in a few hours.

brettwooldridge commented:
@dsyang No clear explanation yet, but removing ``thirdparty/java/kotlin`` entirely, and deleting the dependency in ``thirdparty/java/mapdb/BUCK`` fixed the issue.  ``kotlin-merge`` now runs Kotlin builds against our production code without issue (unit tests may or may not all work, haven't checked them yet).

One possibility is, because there are two classloaders (at least), the main Java app classloader, and the one created by ``JarBackedReflectedKotlinc``, that somehow Kotlin stdlib classes were getting loaded in two separate classloaders.  If that proves to be the cause, then in *the future*, we might need to create a custom classloader for the kotlin compile that is non-delegating -- or at least *delegate last* rather than *delegate first*.

brettwooldridge commented:
I squashed and pushed ``kotlin-compiler2``.

dsyang commented on src/com/facebook/buck/jvm/kotlin/JarBackedReflectedKotlinc.java:
```diff
@@ -0,0 +1,230 @@
+      classLoaderCache.addRef();
+
+      ClassLoader classLoader = classLoaderCache.getClassLoaderForClassPath(
+          this.getClass().getClassLoader(),
```
Found the culprit!

We're running the kotlin compiler with a classpath that includes all the classes loaded for buck.  That's not what we want. The only classes the compiler should know about / have access to are the ones specified in the `compilerClassPath`.

Changing this to `null` allows for successful compilation without needing to change `mapdb` or anything in `third-party`.

dsyang commented on src/com/facebook/buck/jvm/kotlin/KotlinBuckConfig.java:
```diff
@@ -58,67 +63,105 @@ public KotlinBuckConfig(BuckConfig delegate) {
+   *  </ul>
+   *
+   * @return the Kotlin home path
+   */
```
The types of directory structures we consider valid kotlin_home isn't quite clear.  If I have my kotlinc in `~/bin/kotlinc` then this considers `~/bin` to be my kotlin_home which seems like it'd be wrong (`~/bin/lib/kotlin-runtime.jar`? no thanks).

Can we add a verify method that makes sure `kotlin_home`s are either structured like:
```
kotlin_home_as_downloaded_and_unziped_from_website/
    bin/
    lib/
```
or 
```
kotlin_home_as_installed_from_homebrew/
    bin/
    libexec/
        lib/
```

brettwooldridge replied:
@dsyang I added additional tests to verify correct behavior.  The supported structures are...

For _external_ compilation:
 * *external = true* defined in ``.buckconfig``
 * *kotlin_home* pointing directly at directory containing ``kotlinc``
 * *kotlin_home* pointing at directory containing a ``bin`` directory that contains ``kotlinc``
 * No *kotlin_home* defined, ``kotlinc`` in environment ``PATH``

For _in-process_ compilation:
 * *kotlin_home* pointing directly at directory containing ``kotlin-*`` JARs <sup>1</sup>
 * *kotlin_home* pointing at directory containing a ``lib`` directory that contains ``kotlin-*`` JARs
 * *kotlin_home* pointing at directory containing a ``libexec`` directory, that contains a ``lib`` directory, that contains ``kotlin-*`` JARs
 * No *kotlin_home* defined, ``kotlinc`` in environment ``PATH``, assume ``kotlinc`` is in a ``bin`` directory, and the true "kotlin home" is one directory above that.

<sup>1</sup> This supports the case, actually my company's case, where Kotlin jars are present in the absence of an installation structure.  In our case, outside of buck, we essentially download the jars using maven:
```xml
<dependency>
    <groupId>org.jetbrains.kotlin</groupId>
    <artifactId>kotlin-*</artifactId>
    <version>${kotlin-version}</version>
</dependency>
```

kageiit replied:
The usecase mentioned by @brettwooldridge in 1 will be what we would be using for okbuck as well. Thanks!

dsyang replied:
Fair points, didn't think about maven or okBuck.  lets keep it this way then!

dsyang commented on src/com/facebook/buck/jvm/kotlin/Kotlinc.java:
```diff
@@ -0,0 +1,53 @@
+      Path pathToSrcsList);
+
+  String getShortName();
+
```
extra whitespace

brettwooldridge replied:
Removed.

dsyang commented on src/com/facebook/buck/jvm/kotlin/KotlincStep.java:
```diff
@@ -17,69 +17,129 @@
+      String firstOrderStderr = stderr.getContentsAsString(Charsets.UTF_8);
+      Optional<String> returnedStderr;
+      if (declaredDepsBuildResult != 0) {
+        returnedStderr = Optional.of(firstOrderStderr); //processBuildFailure(context, firstOrderStdout, firstOrderStderr);
```
address comment or remove

brettwooldridge replied:
Addressed.

dsyang commented on src/com/facebook/buck/jvm/kotlin/KotlincStep.java:
```diff
@@ -17,69 +17,129 @@
+  @Override
+  public String getDescription(ExecutionContext context) {
+    return getKotlinc().getDescription(
+        ImmutableList.of(), // TODO getOptions(context, getClasspathEntries()),
```
address comment or remove

dsyang commented on src/com/facebook/buck/jvm/kotlin/KotlincToJarStepFactory.java:
```diff
@@ -68,24 +70,41 @@ public void createCompileStep(
             filesystem));
   }
 
+//  @Override
```
address comment or remove

dsyang commented on src/com/facebook/buck/jvm/kotlin/LICENSE:
```diff
@@ -0,0 +1,15 @@
+/*
```
remove

dsyang commented on test/com/facebook/buck/jvm/kotlin/KotlinBuckConfigTest.java:
```diff
@@ -17,168 +17,190 @@
-
-    workspace = TestDataHelper.createProjectWorkspaceForScenario(this, "kotlin_compiler_test", tmp);
-    workspace.setUp();
+    workspace = TestDataHelper
```
I don't think we should create a new testdata directory for this buck config test.  That's pretty heavy when you can just use tmp to create temporary files/directory structures.

Check out how `DBuckConfigTest`,  does it.  We should do it that way and delete the `testdata/faux_kotlin_home` and `testdata/kotlin_compiler_test` directories.

dsyang commented on test/com/facebook/buck/jvm/kotlin/KotlinBuckConfigTest.java:
```diff
@@ -17,168 +17,190 @@
+        .resolve("faux_kotlin_home")
+        .resolve("libexec")
+        .resolve("lib")
+        .resolve("kotlin-stdlib.jar");
```
i think tests are failing because of kotlin-stdlib / kotlin-runtime naming confusion

brettwooldridge replied:
All tests passing now.

brettwooldridge commented:
@dsyang Thanks.  Fixed.  Also cleaned up code, minimized diffs, and fixed all unit tests.  Pending further review, I'm done on my end.

brettwooldridge commented on src/com/facebook/buck/jvm/kotlin/JarBackedReflectedKotlinc.java:
```diff
@@ -0,0 +1,230 @@
+  private static final String INCLUDE_RUNTIME_FLAG = "-include-runtime";
+  private static final String COMPILER_CLASS = "org.jetbrains.kotlin.cli.jvm.K2JVMCompiler";
+  private static final String EXIT_CODE_CLASS = "org.jetbrains.kotlin.cli.common.ExitCode";
+  private static final KotlincVersion VERSION = KotlincVersion.of("in memory");
```
I just basically mimicked what [JarBackedJavac](https://github.com/facebook/buck/blob/c112b5dc9330005cdfeb6c574e5f35b5ed2d6469/src/com/facebook/buck/jvm/java/JarBackedJavac.java#L67) was using.

brettwooldridge commented on src/com/facebook/buck/jvm/kotlin/KotlinBuckConfig.java:
```diff
@@ -44,12 +41,20 @@ public KotlinBuckConfig(BuckConfig delegate) {
+  }
+
+
+  Path getCompilerPath() {
```
Done.

brettwooldridge replied:
Done.

dsyang commented:
@brettwooldridge  I'm seeing a bunch of "done" / "addressed" comments without any changes.  Maybe one of your commits got lost in translation?

brettwooldridge commented:
A bunch?  The only missed ones I found was the change to ``getPathToCompilerBinary()`` and the comment on KotlincStep:117.  Both are pushed now.

dsyang commented:
what about https://github.com/facebook/buck/pull/1283#discussion_r117647080?

Or do you think that belongs in a separate PR?

brettwooldridge commented:
I'm not sure what is particularly "heavy" about the testdata directory, given that it is merely a "structure" and the "binaries" and "jars" themselves are merely files of several dozen bytes.

I have complained before about *truly heavy* testdata contents like that used by the scala tests, but it seems to fall on deaf ears.  [Scala testdata](https://github.com/facebook/buck/tree/master/test/com/facebook/buck/jvm/scala/testdata) has the scala compiler (15MB) and various other scala jars (~10MB) checked in *multiple times* -- totaling over 100MB.  By comparison, I'm not particularly concerned about three folders with fake binary files totaling a 100 bytes or so.

dsyang commented:
By "heavy" I meant cognitive load instead of file size load.  I'm also not worried about the size of those testdata directories.  Perhaps a better way to phrase it is: "Do we really need these to be testdata directories?" 

To me, testdata dirs are useful in places where file content is important (such as KotlinLibraryIntegrationTest where you actually compiling those sources).  In the case of KotlinBuckConfigTest we only care about the existence of those files in that hierarchy. That's exactly what `tmp.newExectuableFile()` and friends provide: the temporary existence of a file without caring about it's contents. 

The "cognitive load" part comes from locality.  To figure out what that test is doing I need to look at the directory structure, then look at the test.  With the way `DBuckConfigTest` does it, you only have to look at the test file to understand what's going on.

brettwooldridge commented:
Done.  I left the ``kotlin_compiler_test`` unchanged as that code was not touched by this PR.

dsyang commented on src/com/facebook/buck/jvm/kotlin/ExternalKotlinc.java:
```diff
@@ -0,0 +1,158 @@
+package com.facebook.buck.jvm.kotlin;
```
https://ci.appveyor.com/project/Facebook/buck/build/3716#L2559

dsyang commented on src/com/facebook/buck/jvm/kotlin/KotlincStep.java:
```diff
@@ -20,66 +20,174 @@
 
-public class KotlincStep extends ShellStep {
-
+@SuppressWarnings("unused")
```
are there warnings left to suppress? I thin we use everything

dsyang commented on test/com/facebook/buck/jvm/kotlin/BUCK:
```diff
@@ -137,8 +137,10 @@ java_library(
 )
 
-standard_java_test(
+java_test(
```
this change should be reverted since we don't have a faux_kotlin_home directory anymore

brettwooldridge commented:
I see you started the merge, does that mean you are going to address the "nits", or do you still want me to change and squash them?  I'm out of the office right now but I could do it when I get in.

dsyang commented:
If you could squash them that'd be great.  I just imported the PR to run it in our internal CI servers in case there are more tests that fail (I don't expect them to)

brettwooldridge commented on src/com/facebook/buck/jvm/kotlin/BUCK:
```diff
@@ -6,6 +6,8 @@ java_immutables_library(
     visibility = ["PUBLIC"],
     deps = [
+        "//src/com/facebook/buck/android:apkmodule",
+        "//src/com/facebook/buck/android:packageable",
```
When I take these out, I get this error:
```
buck/src/com/facebook/buck/jvm/kotlin/KotlinTestDescription.java:93: error: cannot access com.facebook.buck.android.AndroidPackageable
    DefaultJavaLibrary testsLibrary = resolver.addToIndex(defaultJavaLibraryBuilder.build());
                                              ^
  class file for com.facebook.buck.android.AndroidPackageable not found
Errors: 1. Warnings: 0.
```

dsyang commented on src/com/facebook/buck/jvm/kotlin/JarBackedReflectedKotlinc.java:
```diff
@@ -0,0 +1,211 @@
+
+      Object exitCode = compile.invoke(
+          compilerShim,
+          new UncloseablePrintStream(context.getStdErr()),
```
Question: Why are we using a UncloseablePrintStream here? Curious what the problem is with just context.getStdErr().

brettwooldridge replied:
The Kotlin compiler closes the stream and breaks the buck daemon.

dsyang replied:
@brettwooldridge an Infer run is complaining about a [resource leak](http://fbinfer.com/docs/infer-bug-types.html#RESOURCE_LEAK) here:  Any ideas on the best way to fix this?

brettwooldridge commented:
@dsyang That push should fix the warning.

brettwooldridge commented:
Rebased onto master.

brettwooldridge commented:
@kageiit @dsyang Anything else waiting on me here?

dsyang commented:
Don't have my computer with me for the next couple of days will look at it
again on Thursday

On May 30, 2017 3:47 AM, "Brett Wooldridge" <notifications@github.com>
wrote:

> @kageiit <https://github.com/kageiit> @dsyang <https://github.com/dsyang>
> Anything else waiting on me here?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/facebook/buck/pull/1283#issuecomment-304883353>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAgWMV5tN3Z6D6HSDcT7k5k6Ym-cS61Kks5r_B3ogaJpZM4M4A-v>
> .
>

kageiit commented:
@dsyang Ping on this :)

aww seems there are new merge conflicts :( @brettwooldridge

brettwooldridge commented:
@kageiit Jesus, do we really need all these refactors that apparently touch almost every class in the tree?

dsyang commented:
@brettwooldridge @kageiit  We're just waiting on some people from the buck team to do a quick once-over for this PR on our internal phabricator import, unfortunately some people have been OoO so it's not moving as fast as I'd like.  I've been rebasing and updating it internally but I don't know how to make it show up here in github. 

You guys shouldn't need to do anything more here.  It's my job now to push someone internally to look at it :/

brettwooldridge commented:
@dsyang @kageiit Just checking on this.  It is closed, but not merged?  A little concerned, because pull #1358 seems based off of the existing mainline but will surely create conficts if merged before this.

Coneko commented:
It was merged as 800d5a9669417d5a8b4bc455e34a629422ed1ff1.

sdwilsh commented:
"merge" means someone actually used the merge button in the GitHub UI. 
That will never happen with this project, so PRs always show up as "closed".

brettwooldridge commented:
@sdwilsh Sorry, I missed the ``facebook-github-bot closed this in 800d5a9 9 days ago`` message.  I guess the next question is, when might we see this is a release?

sdwilsh commented:
It looks like @aiked has been doing those for Buck.

dsyang commented:
@brettwooldridge  if you don't want to wait for a release you could also use a .buckversion file https://buckbuild.com/concept/buckversion.html
